pkg/controller/common/wrappedqueue.go

```diff
@@ -0,0 +1,276 @@
+package common
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+type errDropFromQueue struct{}
+
+func (e *errDropFromQueue) Error() string {
+	return fmt.Sprint("drop from queue")
+}
+
+type errNotPathologicalFailure struct{}
+
+func (e *errNotPathologicalFailure) Error() string {
+	return fmt.Sprintf("no pathological failure")
+}
+
+// Any function which should not pathologically fail should return this error
+// or have it someplace within the returned error chain.
+var ErrPathological error = &errNotPathologicalFailure{}
+
+// Any function that should stop execution and be dropped from the queue should
+// return this error or have it someplace within the returned error chain.
+var ErrDropFromQueue error = &errDropFromQueue{}
+
+// Functions are not hashable, which is a core requirement of the workqueue
+// interface. However, structs are. So WrappedQueue attaches functions to an
+// instance of this struct before adding them into the queue.
+type executable struct {
+	name string
+	run  func() error
+}
+
+// WrappedQueue serves the following purposes:
+// 1. Abstract the minutae of setting up and managing a workqueue away from a
+// controller.
+// 2. Enable having multiple work queues per controller.
+// 3. Enables the addition of functions onto the work queue instead of just
```

Comment:
non-blocking praise:
This was much needed!

Reply:
Thanks! Something I'd like to point out about this change is that prior to this, the workqueues were just having kube objects pushed onto and pulled from them with no context (such as, was it added, changed, deleted, etc.) about why the object was pushed onto the queue. Additionally, workqueues are untyped, which means you have to do type switching if your controller deals with many different object types and events. There is now a `TypedRateLimitingQueue` where you don't have to do that: https://github.com/kubernetes/client-go/blob/master/util/workqueue/rate_limiting_queue.go#L73-L79 (though we'd have to wait until after the next Kube rebase to be able to use it).

Ideally, one would have a separate workqueue for each type which is initially what I was going to do, but I can also see why this would be undesirable since you could easily end up with race conditions if you're not careful. While `TypedRateLimitingQueue` eliminates the need to do type switches, it does not implement the worker loop, nor does it solve the "what should I do with this object now?". `WrappedQueue` solves both of those problems by implementing the worker loop and exposing an API to push functions onto the queue instead of objects. In my opinion, this ends up being more straightforward since the caller just has to enqueue a function for execution.

Reply:
FYI, once https://github.com/openshift/machine-config-operator/pull/4629 lands, all our controllers will begin to use `TypedRateLimitingQueue`, the old ones are deprecated and causes verify errors (: